webdav/webdav.go

```diff
@@ -96,45 +96,116 @@ func (h *Handler) lock(now time.Time, root string) (token string, status int, er
+	// 2. If any locks *are* held on src/dst, check that the If header satisfies them.
+	// 3. As part of checking #2, validate any other constraints implied in the If header.
+
 	hdr := r.Header.Get("If")
-	if hdr == "" {
-		// An empty If header means that the client hasn't previously created locks.
-		// Even if this client doesn't care about locks, we still need to check that
-		// the resources aren't locked by another client, so we create temporary
-		// locks that would conflict with another client's locks. These temporary
-		// locks are unlocked at the end of the HTTP request.
-		now, srcToken, dstToken := time.Now(), "", ""
-		if src != "" {
-			srcToken, status, err = h.lock(now, src)
-			if err != nil {
-				return nil, status, err
-			}
+	ih := ifHeader{}
+	// Parse the If header first, since if it's bad we should just fail out, even if we
+	// don't need it.
+	if hdr != "" {
+		var ok bool
+		ih, ok = parseIfHeader(hdr)
+		if !ok {
+			return nil, http.StatusBadRequest, errInvalidIfHeader
 		}
-		if dst != "" {
-			dstToken, status, err = h.lock(now, dst)
-			if err != nil {
-				if srcToken != "" {
-					h.LockSystem.Unlock(now, srcToken)
-				}
-				return nil, status, err
-			}
+	}
+
+	// Even if the client hasn't previously created locks, another principle may have.
+	// So, we still need to check that the resources aren't locked by another client.
+	// To do so, we create temporary locks on the requested resources. If the resoure is
+	// locked, the operation will fail. In that case, we will check if the request
+	// presented a lock for that particular resource.
+
+	// Any temporary locks are removed at the end of the request.
+	now, srcToken, dstToken := time.Now(), "", ""
+	if src != "" {
+		srcToken, status, err = h.speculativeLock(now, src, hdr)
+		if err != nil {
+			return nil, status, err
 		}
+	}
 
-		return func() {
-			if dstToken != "" {
-				h.LockSystem.Unlock(now, dstToken)
-			}
+	if dst != "" {
+		dstToken, status, err = h.speculativeLock(now, dst, hdr)
+		if err != nil {
 			if srcToken != "" {
 				h.LockSystem.Unlock(now, srcToken)
 			}
-		}, 0, nil
+			return nil, status, err
+		}
+	}
+
+	speculativeLockRelease := func() {
+		if dstToken != "" {
+			h.LockSystem.Unlock(now, dstToken)
+		}
+		if srcToken != "" {
+			h.LockSystem.Unlock(now, srcToken)
+		}
+	}
+
+	// Exclude src/dst from the lock search if we already have a lock for them.
+	if srcToken != "" {
+		src = ""
+	}
+
+	if dstToken != "" {
+		dst = ""
+	}
+
+	if src == "" && dst == "" {
+
+		if len(ih.lists) == 0 {
+			// No conditions to evaluate and no src/dst constraints to check.
+			// Everything with the request is good. Return success.
+			return speculativeLockRelease, 0, nil
+		}
+
+		// In this case, we have created temporary locks on any resource we care about.
+		// This means that none of the conditions can evaluate to true. Fall through with
+		// an empty list to ensure consistent error handling
+		ih.lists = []ifList{}
```

Comment:
I'm not a huge fan of this -- it feels a little messy to depend on the behaviour of atLeastOnIfListPasses, but I think the newly created functions name + the comment make it clear enough what's expected.